### PR TITLE
feat: prove blocks using reth

### PR DESCRIPTION
**Pull request by Brechtpd**
- currently on reth v1.0.0-rc.2
- state root verified outside reth for now

reth v1.0.0-rc.2 taiko changes: https://github.com/taikoxyz/taiko-reth/pull/6
revm v36 taiko changes: https://github.com/taikoxyz/revm/pull/10

### CODE REVIEW

**Comment by CeciliaZ030** [comment]
lgtm
shouldn't be too difficult to merge the Proof of Equivalence PR

File: provers/risc0/guest/src/mem.rs [hunk 1]
```diff
@@ -33,3 +33,8 @@ pub unsafe extern "C" fn calloc(nobj: usize, size: usize) -> *mut c_void {
 pub unsafe extern "C" fn free(_size: *const c_void) {
     // Intentionally a no-op, since the zkvm allocator is a bump allocator
 }
+
+#[no_mangle]
+pub extern "C" fn __ctzsi2(x: u32) -> u32 {
```
Comment (smtmfft) [inline provers/risc0/guest/src/mem.rs:38]: why we need this one?

**Review by smtmfft — APPROVED** [review]
LGTM!
just wonder if that incorrect access list gas issue does not exist under this reth model?

File: host/config/chain_spec_list_default.json [hunk 2]
```diff
@@ -127,33 +127,5 @@
         "genesis_time": 0,
         "seconds_per_slot": 1,
         "is_taiko": true
-    },
-    {
-        "name": "taiko_mainnet",
```
[the remark this answers is not in the document]
Comment (smtmfft) [reply host/config/chain_spec_list_default.json:132]: I guess it is moved to taiko-reth?? then why we leave the others?

File: core/Cargo.toml [hunk 3]
```diff
@@ -13,6 +13,13 @@ sgx-prover = { path = "../provers/sgx/prover", optional = true }
 # raiko
 raiko-lib = { workspace = true, features = ["c-kzg"] }
 
+# reth
+reth-primitives.workspace = true
```
[the remark this answers is not in the document]
Comment (Brechtpd) [reply core/Cargo.toml:17]: Ah reth is doing `package.workspace = true` everywhere now, so I got in the habit of doing the same. Though they are rust maxis in reth, so should be good practice? If so, we probably want to use this style everywhere in raiko as well.

File: provers/risc0/guest/src/mem.rs (hunk 1 above)
Comment (Brechtpd) [reply provers/risc0/guest/src/mem.rs:38]: Some issue with compiling secp256k1 if I remember correctly where it's undefined similar to the malloc stuff, but I think Cecilia's stuff will change this again so didn't give it too much thought for now, I think likely the code path isn't even triggered but not sure.

**Comment by Brechtpd** [comment]
> LGTM! just wonder if that incorrect access list gas issue does not exist under this reth model?

Still necessary! https://github.com/taikoxyz/taiko-reth/blob/d20c700ca296f7e4227b4c8c254a8d0f948bbd36/crates/ethereum/evm/src/execute.rs#L222

**Comment by smtmfft** [comment]
> > LGTM! just wonder if that incorrect access list gas issue does not exist under this reth model?
> 
> Still necessary! https://github.com/taikoxyz/taiko-reth/blob/d20c700ca296f7e4227b4c8c254a8d0f948bbd36/crates/ethereum/evm/src/execute.rs#L222

Weird..is that a bug existed in both zeth & official reth?? I thought reth should be ok as its target is geth equivalence, should not have such a bug.

**Comment by Brechtpd** [comment]
Blocks don't skip invalid transactions on Ethereum, so the behavior we're triggering is taiko specific, so not a bug there.

File: provers/sgx/prover/src/sgx_register_utils.rs [hunk 4]
```diff
@@ -276,24 +274,20 @@ pub async fn register_sgx_instance(
...
-    let mut wallet: alloy_signer_wallet::LocalWallet = sender_priv_key.as_str().parse().unwrap();
+    let mut wallet: PrivateKeySigner = sender_priv_key.as_str().parse().unwrap();
     wallet.set_chain_id(Some(chain_id));
     println!("wallet: {:?}", wallet);
 
     // init rpc conn
     let http = Http::new(Url::parse(l1_rpc_url).expect("invalid rpc url"));
     let provider = ProviderBuilder::new()
         .with_recommended_fillers()
-        .signer(EthereumSigner::from(wallet.clone()))
```
Comment (johntaiko) [inline provers/sgx/prover/src/sgx_register_utils.rs:287]: Can't send transaction without this

**Pull request merged**